connector/src/main/java/com/google/cloud/bigquery/connector/common/IteratorMultiplexer.java

```diff
@@ -96,46 +96,50 @@ void readAhead() {
   }
 
   private class QueueIterator<T> implements Iterator<T> {
-    private final ArrayBlockingQueue<T> queue = new ArrayBlockingQueue<>(/*capacity=*/ 1);
-    private final AtomicBoolean done = new AtomicBoolean(false);
+    private final ArrayBlockingQueue<Object> queue = new ArrayBlockingQueue<>(/*capacity=*/ 2);
+    private final Semaphore sem = new Semaphore(1);
 
-    private T t = null;
+    private Object t = null;
 
     @Override
     public boolean hasNext() {
-      if (!mightHaveNext()) {
+      if (t == TERMINAL_SENTINEL) {
         return false;
       }
-      t = null;
       try {
-        while (t == null && mightHaveNext()) {
-          t = queue.poll(10, TimeUnit.MILLISECONDS);
-        }
+        t = queue.take();
+        sem.release();
       } catch (InterruptedException e) {
-        done.set(true);
         // We expect all iterators to either make progress together or finish.
         // This starts the cleanup process to halt all workers.
         worker.interrupt();
-        throw new RuntimeException(e);
-      }
-      if (t == null) {
-        done.set(true);
-        return false;
+        t = TERMINAL_SENTINEL;
       }
-      return true;
-    }
-
-    private boolean mightHaveNext() {
-      return !done.get() || !queue.isEmpty();
+      return t != TERMINAL_SENTINEL;
     }
 
     @Override
     public T next() {
-      Preconditions.checkState(t != null, "next element cannot be null");
-      if (rethrow != null) {
-        throw rethrow;
+      Preconditions.checkState(t != TERMINAL_SENTINEL, "next element cannot be null");
```

Comment:
Error message should be "next element cannot be the sentinel"

Reply:
changed to "no next message"